File: library/src/main/java/com/yy/codex/uikit/UIBarButtonItem.java
package com.yy.codex.uikit;

import android.content.Context;

import com.yy.codex.foundation.NSInvocation;

/**
 * Created by cuiminghui on 2017/1/18.
 */

public class UIBarButtonItem extends UIBarItem {

    boolean mSystemBackItem = false;

    public boolean isSystemBackItem() {
        return mSystemBackItem;
    }

    private Object mTarget;
    private String mAction;

    public UIBarButtonItem() {

    }

    public UIBarButtonItem(UIImage image, Object target, String action) {
        mImage = image;
        mTarget = target;
        mAction = action;
    }

    public UIBarButtonItem(String title, Object target, String action) {
        mTitle = title;
        mTarget = target;
        mAction = action;
    }

    protected UIView mCustomView = null;

    public UIView getCustomView() {
        return mCustomView;
    }

    public void setCustomView(UIView customView) {
        this.mCustomView = customView;
    }

    protected double mWidth = 0.0;

    public double getWidth() {
        return mWidth;
    }

    public void setWidth(double width) {
        this.mWidth = width;
    }

    protected UIEdgeInsets mInsets;

    public UIEdgeInsets getInsets() {
        if (mInsets == null) {
            mInsets = new UIEdgeInsets(0, 8, 0, 8);
        }
        return mInsets;
    }

    public void setInsets(UIEdgeInsets insets) {
        mInsets = insets;
    }

    protected UIEdgeInsets mImageInsets;

    @Override
    public UIView getContentView(Context context) {
        if (mCustomView != null) {
            if (mWidth > 0.0) {
                mCustomView.setFrame(new CGRect(0, 0, mWidth, Math.min(44.0, mCustomView.getFrame().getHeight())));
            }
            mCustomView.setMarginInsets(getInsets());
            return mCustomView;
        }
        else if (mView == null) {
            UIButton button = new UIButton(context);
            if (mTitle != null) {
                button.setFont(new UIFont(17));
                button.setTitle(mTitle, UIControl.State.Normal);
            }
            if (mImage != null) {
                button.setImage(mImage, UIControl.State.Normal);
            }
            button.setFrame(new CGRect(0, 0, button.intrinsicContentSize().getWidth(), 44));
            button.setImageEdgeInsets(getImageInsets());
            if (mTarget != null && mAction != null) {
                button.addTarget(mTarget, mAction, UIControl.Event.TouchUpInside);
            }
            mView = button;
            mView.setMarginInsets(getInsets());
        }
        return super.getContentView(context);
    }

}

File: library/src/main/java/com/yy/codex/uikit/UIBarButtonItem.kt
package com.yy.codex.uikit

import android.content.Context

import com.yy.codex.foundation.NSInvocation

/**
 * Created by cuiminghui on 2017/1/18.
 */

class UIBarButtonItem(val target: Any?, val action: String?) : UIBarItem() {

    var isSystemBackItem = false
        internal set
    var customView: UIView? = null
    var width = 0.0
    var insets: UIEdgeInsets = UIEdgeInsets(0.0, 8.0, 0.0, 8.0)

    constructor(title: String?, target: Any?, action: String?) : this(target, action) {
        this.title = title
    }

    constructor(image: UIImage?, target: Any?, action: String?) : this(target, action) {
        this.image = image
    }

    override fun getContentView(context: Context): UIView? {
        if (customView != null) {
            if (width > 0.0) {
                customView?.frame = CGRect(0.0, 0.0, width, Math.min(44.0, (customView?.frame?.height ?: 0.0)))
            }
            customView?.marginInsets = insets
            return customView
        }
        else if (mView == null) {
            val button = UIButton(context)
            title?.let {
                button.setFont(UIFont(17f))
                button.setTitle(it, UIControl.State.Normal)
            }
            image?.let {
                button.setImage(it, UIControl.State.Normal)
            }
            button.frame = CGRect(0.0, 0.0, button.intrinsicContentSize().width, 44.0)
            button.setImageEdgeInsets(imageInsets)
            if (target != null && action != null) {
                button.addTarget(target!!, action!!, UIControl.Event.TouchUpInside)
            }
            mView = button
            mView?.marginInsets = insets
        }
        return super.getContentView(context)
    }

}
